feat: share IPFS node between browser tabs (#3081)
ipfs/js-ipfs: packages/ipfs-message-port-server/README.md

# ipfs-message-port-server <!-- omit in toc -->

[![](https://img.shields.io/badge/made%20by-Protocol%20Labs-blue.svg?style=flat-square)](http://protocol.ai)
[![](https://img.shields.io/badge/project-IPFS-blue.svg?style=flat-square)](http://ipfs.io/)
[![](https://img.shields.io/badge/freenode-%23ipfs-blue.svg?style=flat-square)](http://webchat.freenode.net/?channels=%23ipfs)
[![Travis CI](https://flat.badgen.net/travis/ipfs/js-ipfs)](https://travis-ci.com/ipfs/js-ipfs)
[![Codecov branch](https://img.shields.io/codecov/c/github/ipfs/js-ipfs/master.svg?style=flat-square)](https://codecov.io/gh/ipfs/js-ipfs)
[![Dependency Status](https://david-dm.org/ipfs/js-ipfs/status.svg?path=packages/ipfs-message-port-server)](https://david-dm.org/ipfs/js-ipfs?path=packages/ipfs-message-port-server)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/feross/standard)

> A library for providing IPFS node over [message channel][]. This library enables

> applications running in the different JS context to use [IPFS API](https://github.com/ipfs/js-ipfs/tree/master/docs/api) (subset) via `ipfs-message-port-client`.

Reviewer: ```suggestion
applications running in the different JS context to use [IPFS API](https://github.com/ipfs/js-ipfs/tree/master/docs/core-api) (subset) via `ipfs-message-port-client`.
```

## Lead Maintainer <!-- omit in toc -->

[Alex Potsides](https://github.com/achingbrain)

## Table of Contentens <!-- omit in toc -->

- [Install](#install)
- [Usage](#usage)
- [Notes on Performance](#notes-on-performance)
- [Contribute](#contribute)
- [License](#license)

## Install

```bash
$ npm install --save ipfs-message-port-server
```

## Usage

> This library can wrap JS IPFS node and expose it over the [message channel][].

Reviewer 2: ```suggestion
This library can wrap a JS IPFS node and expose it over the [message channel][].
```

It assumes `ipfs-message-port-client` on the other end, however it is not

> strictly necessary anything compling with wire protocol will do.

Reviewer 2: ```suggestion
strictly necessary anything complying with the wire protocol will do.
```

It provides following API subset:

- [`ipfs.dag`](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/DAG.md)
- [`ipfs.block`](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/BLOCK.md)
- [`ipfs.add`](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/FILES.md#ipfsadddata-options)
- [`ipfs.cat`](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/FILES.md#ipfscatipfspath-options)
- [`ipfs.files.stat`](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/FILES.md#ipfsfilesstatpath-options)

> Server is designed to run in the [SharedWorker][] (although it is possible to

Reviewer 2: ```suggestion
The server is designed to run in a [SharedWorker][] (although it is possible to
```

> run it in the other JS contexts). Example below illustrates running js-ipfs

Reviewer 2: ```suggestion
run it in the other JS contexts). The example below illustrates running a js-ipfs
```

> node in [SharedWorker][] and exposing it to all connected ports

Reviewer 2: ```suggestion
node in a [SharedWorker][] and exposing it to all connected ports
```

const IPFS = require('ipfs')
const { IPFSService } = require('ipfs-message-port-server')
const { Server } = require('ipfs-message-port-server/src/server')

const main = async () => {
  const ports = []
  // queue connections that occur while node was starting.

>   self.onconnect = ({ports}) => connections.push(...ports)

Author: Oh that was an error, it meant to push into `ports` above. Thanks for catching I'll fix it.

  const ipfs = await IPFS.create()
  const service = new IPFSService(ipfs)
  const server = new Server(service)

  // connect new ports and queued ports with the server.
  self.onconnect = ({ports}) => server.connect(ports[0])
  for (const port of ports.splice(0)) {
    server.connect(port)
  }
}

main()
```

> ### Additional Consideration

Reviewer: This section should be in TOC, for easy discovery when skimming (similar note in `README` for `-client`)
```suggestion
### Performance Consideration
```

### Notes on Performance

> Since the data over [message channel][] is copied via

Reviewer 2: ```suggestion
Since the data sent over the [message channel][] is copied via
```

> [structured cloning algorithm][] it may lead to suboptimal

Reviewer 2: ```suggestion
the [structured cloning algorithm][] it may lead to suboptimal
```

> results (espacially with large binary data). In order to avoid unecessary

Reviewer 2: ```suggestion
results (especially with large binary data). In order to avoid unnecessary
```

> copying server will transfer all the [Transferable][] which will be emptied

Reviewer 2: ```suggestion
copying the server will transfer all passed [Transferable][]s which will be emptied
```

on the server side. This should not be a problem in general as IPFS node itself
does not retain references to returned values, but is something to keep in mind

> when doig something custom.

Reviewer 2: ```suggestion
when doing something custom.
```
